File: sdks/python/memexai/_paths.py
from typing import Optional
from .errors import MemexError
from .types import RequestContext

def validate_virtual_path(path: str) -> None:
    if not path or not isinstance(path, str):
        raise MemexError("INVALID_PATH", "path is required")

    if path.startswith("/") or "\\" in path or "//" in path:
        raise MemexError("INVALID_PATH", "Path must be a relative slash-delimited path")

    segments = path.split("/")
    if any(segment in ("", ".", "..") for segment in segments):
        raise MemexError("INVALID_PATH", "Path cannot contain empty, dot, or dot-dot segments")

    if segments[0] == "users":
        raise MemexError("PHYSICAL_PATH_FORBIDDEN", "Agents cannot use physical users/{userId} paths")

def virtual_to_physical(path: str, ctx: RequestContext) -> str:
    validate_virtual_path(path)

    if path == "user" or path.startswith("user/"):
        if not ctx.user_id:
            raise MemexError("USER_ID_REQUIRED", "userId is required for user/** memory paths")
        if path == "user":
            return f"users/{ctx.user_id}"
        return f"users/{ctx.user_id}/{path[len('user/'):]}"

    if path == "shared" or path.startswith("shared/"):
        return path

    raise MemexError("UNKNOWN_MOUNT", "Path must start with user/ or shared/")

def physical_to_virtual(physical_path: str, ctx: RequestContext) -> Optional[str]:
    user_prefix = f"users/{ctx.user_id}/"
    if physical_path.startswith(user_prefix):
        return f"user/{physical_path[len(user_prefix):]}"

    if physical_path == f"users/{ctx.user_id}":
        return "user"

    if physical_path == "shared" or physical_path.startswith("shared/"):
        return physical_path

    return None

def assert_writable_virtual_path(path: str) -> None:
    validate_virtual_path(path)
    if not (path == "user" or path.startswith("user/")):
        raise MemexError("READ_ONLY_MOUNT", "Agents cannot write to shared/**")

def prefix_to_physical(prefix: Optional[str], ctx: RequestContext) -> Optional[str]:
    if not prefix:
        return None
    normalized = prefix[:-1] if prefix.endswith("/") else prefix
    return virtual_to_physical(normalized, ctx)

File: packages/core/src/paths.ts
import { MemexError } from "./errors"

export type ToolContext = {
  userId: string
  actor?: string
  toolCallId?: string
}

export type SharedWriteMode = "read_only" | "rw"

export type MemoryPermissions = {
  sharedWriteMode: SharedWriteMode
  writableMounts: string[]
  readOnlyMounts: string[]
}

export function resolveMemoryPermissions(input: { sharedWriteMode?: SharedWriteMode } = {}): MemoryPermissions {
  const sharedWriteMode = input.sharedWriteMode ?? "read_only"
  return {
    sharedWriteMode,
    writableMounts: sharedWriteMode === "rw" ? ["user", "shared"] : ["user"],
    readOnlyMounts: sharedWriteMode === "rw" ? [] : ["shared"],
  }
}

export function validateVirtualPath(path: string): void {
  if (!path || typeof path !== "string") {
    throw new MemexError("INVALID_PATH", "path is required")
  }

  if (path.startsWith("/") || path.includes("\\") || path.includes("//")) {
    throw new MemexError("INVALID_PATH", "Path must be a relative slash-delimited path")
  }

  const segments = path.split("/")
  if (segments.some((segment) => segment === "" || segment === "." || segment === "..")) {
    throw new MemexError("INVALID_PATH", "Path cannot contain empty, dot, or dot-dot segments")
  }

  if (segments[0] === "users") {
    throw new MemexError("PHYSICAL_PATH_FORBIDDEN", "Agents cannot use physical users/{userId} paths")
  }
}

export function virtualToPhysical(path: string, ctx: ToolContext): string {
  validateVirtualPath(path)

  if (path === "user" || path.startsWith("user/")) {
    if (!ctx.userId) {
      throw new MemexError("USER_ID_REQUIRED", "userId is required for user/** memory paths")
    }
    return path === "user" ? `users/${ctx.userId}` : `users/${ctx.userId}/${path.slice("user/".length)}`
  }

  if (path === "shared" || path.startsWith("shared/")) {
    return path
  }

  throw new MemexError("UNKNOWN_MOUNT", "Path must start with user/ or shared/")
}

export function physicalToVirtual(physicalPath: string, ctx: ToolContext): string | null {
  if (physicalPath.startsWith(`users/${ctx.userId}/`)) {
    return `user/${physicalPath.slice(`users/${ctx.userId}/`.length)}`
  }

  if (physicalPath === `users/${ctx.userId}`) {
    return "user"
  }

  if (physicalPath === "shared" || physicalPath.startsWith("shared/")) {
    return physicalPath
  }

  return null
}

export function assertWritableVirtualPath(path: string, permissions: MemoryPermissions = resolveMemoryPermissions()): void {
  validateVirtualPath(path)
  if (path === "user" || path.startsWith("user/")) return
  if ((path === "shared" || path.startsWith("shared/")) && permissions.writableMounts.includes("shared")) return
  if (path === "shared" || path.startsWith("shared/")) {
    throw new MemexError("READ_ONLY_MOUNT", "Agents cannot write to shared/**")
  }
  throw new MemexError("UNKNOWN_MOUNT", "Path must start with user/ or shared/")
}

export function prefixToPhysical(prefix: string | undefined, ctx: ToolContext): string | null {
  if (!prefix) return null
  const normalized = prefix.endsWith("/") ? prefix.slice(0, -1) : prefix
  return virtualToPhysical(normalized, ctx)
}
